### ISSUE TITLE
Secret consumption - External Secrets

### ISSUE THREAD

**Issue by KenBakerTGT** [go-vela/community#140]
## Description
Requesting a more flexible integration of external secrets. At a step level in a pipeline would like to 'map' a secret to the externally sourced secret. 

An example of how this could work:

## Use secrets plugin to read my external secrets and load to path vela/secrets/my_secrets
secrets:
  - name: vault_username
  - name: vault_password
  
  - origin:
      name: vault
      image: target/secret-vault
      pull: always
      secrets: [ vault_username, vault_password ]
      parameters:
        items:
          - source: secret/vela/my_secrets_in_vault
            path: my_secrets

### Run a step using a custom image, but pass the external secrets to it  as env variables using similar secret mapping. 
steps:
  - name: my_step_to_run_custom_image
    image: my_custom_image
+   secrets: 
         - source: vela/secrets/my_secrets/secret_name     ***this would be the path to where my external secret is stored
            target:  secret_name
    commands:
        - some command  that uses my secret "print $SECRET_NAME"
 
* Why is this important?
This has a few benefits. 
1. When I build custom images to run in a pipeline, if I can map an externally sourced secret to an env variable in the container, I no longer need to put code in my image to read those locations.  Nor do I need to communicate to consumers of my image the details around where these need to be stored(i.e. the path)
2. If I have secrets in an external vault - I can more easily reuse them without having to map in different source/path combinations in the secret plugin.  
* Will this make something better, faster, etc?
This has potential of making integration between secrets and usage in steps simple. I can map any secret to any step for usage, no need to add additional sources in secrets plugin or understand what path the plugin may require. 
* Does this have an impact on security?
I don't see any impact on security
-->

**Comment by colindean** [comment]
Throwing in my vote here. I'm about to migrate a dozen+ repos to use external secrets instead of internal and the inability to easily route the external secret into an environment variable is surprising. I'm trying to plan a workaround but we have a few custom templates that assume that secrets will be available at certain environment variables. It's unclear how best to populate those environment variables from the file with the secret.

**Comment by colindean** [comment]
We decided against using external secrets for now because doing so would have required significant modifications to our templates, which expect environment variables containing the secrets they consume. We'll revisit when external secrets supports environment variable placement or when we have time to expand our templates to support file-based secrets.

**Comment by benzvan** [comment]
One possible fix for this would be to create a `.env` file (or `/secrets/.env` or whatever) that is loaded into the container runtime for every step, making external secrets available as environment variables like internal secrets.

It would have to be:
* Immutable
* created by vela
* marked as secrets to prevent echoing to the logs

**Comment by benzvan** [comment]
@colindean for tasks where we set commands inside the vela task, we use an `export var=$(cat /secret/file)` step to read the secret into a variable. For other tasks it's extremely difficult unless the container has configuration to allow it to read secrets from a file. (Now that I think about it, having a `secret` format in the step to allow a secret to be read from a file would also work.)

**Comment by colindean** [comment]
I ran into this again while considering using external secrets again.

This time, I ~have to use external secrets instead of falling back to internal secrets because of the number of secrets that need to be managed across multiple organizations.

**Comment by benzvan** [comment]
@colindean Yes. It's probably best not to use an internal secret store in a CI environment just for security purposes, so the easier we can make integrating with an external secret store the better. The `export` solution we're using above is only helpful when we control the `commands` run by the plugin and configuring the plugin to read from a file is only helpful when we wrote the plugin ourselves.
